### PR TITLE
Implement c10::Tensor

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#14819 Implement c10::Tensor**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13348039/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #14820 Convert caffe2/aten Tensors to/from c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13348044/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15195 Use C10Tensor in the dispatcher&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13461640/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15324 Fix C10_API/C10_EXPORT for op schema registration&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13500679/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15367 Update flat_hash_map&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13510912/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15199 Move LayerNorm op schema to c10&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13462124/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15243 Enable calling caffe2 LayerNorm from PyTorch and JIT&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13473791/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15312 Remove Context from c10 operator schemas&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13495323/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15316 Move files to/from c10/core and c10/util&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13498493/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15317 Clean up Half&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13498492/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15407 caffe2::Tensor::is_same()&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D13520389/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15853 Move blob to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605256/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15854 Code style cleanup&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605258/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15855 Remove some dependencies from ivalue.h to ATen&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605259/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #15856 Fix export macros&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D13605257/)

This is a minimal wrapper for a c10::TensorImpl,
maybe destined for greatness later when we move caffe2::Tensor or at::Tensor into c10.

Differential Revision: [D13348039](https://our.internmc.facebook.com/intern/diff/D13348039/)

### CODE REVIEW

File: c10/core/Tensor.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <c10/core/TensorImpl.h>
+#include <c10/core/UndefinedTensorImpl.h>
+#include <c10/macros/Macros.h>
+
+namespace c10 {
+
+class C10Tensor final {
```
[the remark this answers is not in the document]
Comment (ezyang) [reply c10/core/Tensor.h]: yes please!

File: c10/core/Tensor.cpp [hunk 2]
```diff
@@ -0,0 +1 @@
+#include <c10/core/Tensor.h>
```
Comment (ezyang) [inline c10/core/Tensor.cpp]: Didn't we agree not to add empty cpp files?

File: c10/core/Tensor.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (smessmer) [reply c10/core/Tensor.h]: I added a comment.

File: c10/core/Tensor.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  using TensorImplPtr = intrusive_ptr<TensorImpl, UndefinedTensorImpl>;
+public:
+  explicit C10Tensor(TensorImplPtr impl) noexcept;
+
+  C10Tensor(const C10Tensor&) = default;
+  C10Tensor(C10Tensor&&) noexcept = default;
+  C10Tensor& operator=(const C10Tensor&) = default;
+  C10Tensor& operator=(C10Tensor&&) noexcept = default;
+
+  const TensorImplPtr &impl() const & noexcept;
```
[the remark this answers is not in the document]
Comment (smessmer) [reply c10/core/Tensor.h]: This is not the equivalent to the `TensorImpl* at::Tensor::getUnsafeTensorImpl`, but the equivalent to the `intrusive_ptr<TensorImpl> at::getTensorImpl`. So it is safe. This is actually the recommended way for kernels on non-standard backends (i.e. backends that have their own TensorImpl class) to get a hold of said TensorImpl.

[the remark this answers is not in the document]
Comment (ezyang) [reply c10/core/Tensor.h]: Now that Greg has reminded me about it, I *would* quite like it if this method was named something easier to grep for, and made it more obvious that it is internal (so sure, not unsafe, but internal; shouldn't be used by most people). internalGetTensorImpl?

**Review by ezyang — APPROVED** [review]
Since we have clangr, it is not too difficult to rename impl post facto, so I'll accept this to unblock. But I don't see the code comment for C10Tensor anywhere in this PR still...

File: c10/core/Tensor.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (smessmer) [reply c10/core/Tensor.h]: Isn't "internal" and "impl" conveying the same thing?

[the remark this answers is not in the document]
Comment (ezyang) [reply c10/core/Tensor.h]: The naming discipline I like to use is, if something should only be called rarely and by people who know what it does, it should have a long and clunky name. Short names for things we want people to use frequently.

**Pull request closed without merging**